src/mempool.cpp

```diff
@@ -0,0 +1,96 @@
+#include "veriblock/mempool.hpp"
+
+#include "veriblock/stateless_validation.hpp"
+
+namespace altintegration {
+
+bool MemPool::submitATV(const std::vector<ATV>& atvs,
+                        const AltTree& tree,
+                        ValidationState& state) {
+  for (size_t i = 0; i < atvs.size(); ++i) {
+    if (!checkATV(atvs[i], state, tree.getParams(), tree.vbk().getParams())) {
+      return state.Invalid("mempool-submit-atv", i);
+    }
+
+    stored_atvs.push_back(atvs[i]);
+  }
+
+  return true;
+}
+
+bool MemPool::submitVTB(const std::vector<VTB>& vtbs,
+                        const AltTree& tree,
+                        ValidationState& state) {
+  for (size_t i = 0; i < vtbs.size(); ++i) {
+    if (!checkVTB(
+            vtbs[i], state, tree.vbk().getParams(), tree.btc().getParams())) {
+      state.Invalid("mempool-submit-vtb", i);
+    }
+
+    stored_vtbs.push_back(vtbs[i]);
+  }
+
+  return true;
+}
+
+bool MemPool::getPop(const AltBlock& current_block,
+                     AltTree& tree,
+                     AltPopTx* out_data,
+                     ValidationState& state) {
+  static constexpr size_t vbk_prev_block_hash_size =
+      decltype(VbkBlock::previousBlock)::size();
+
+  if (!tree.setState(current_block.getHash(), state)) {
+    return state.Invalid("mempool-get-pop");
+  }
+
+  for (const auto& atv : stored_atvs) {
+    VbkBlock first_block =
+        atv.context.size() != 0 ? atv.context[0] : atv.containingBlock;
```

Comment:
```suggestion
        !atv.context.empty()? atv.context[0] : atv.containingBlock;
```

Reply:
Better to reverse comparison:
atv.context.empty()? atv.containingBlock : atv.context[0];